Adjust std.conv.parse to accept underscores in numeric strings

McSherry opened the pull request:
Without this pull request:
```D
auto s = "FF_FF";
auto n = parse!int(s, 16);
writeln(n);  // 255
```

With this pull request:
```D
auto s = "FF_FF";
auto n = parse!int(s, 16);
writeln(n);  // 65535
```

Suggested in #3374.
See also: #3375.

schveiguy commented:
I think it needs to be optional, because this alters behavior of an existing function. Not sure the right API for making it optional, though.

McSherry commented:
A third parameter with a default value would work, would it not?

```D
alias AllowUnderscores = Flag!"AllowUnderscores";

Target parse(Target, Source)(ref Source s, uint radix, AllowUnderscores us = AllowUnderscores.no)
```

McSherry commented:
I added underscores being optional as overloads rather than a parameter with a default value to keep it below the 80-character soft limit.

I also added some documentation for the overload of `parse` taking a radix so that it stands out a bit more.

schveiguy commented:
> I added underscores being optional as overloads rather than a parameter with a default value to keep it below the 80-character soft limit.

I think it's better to use the default parameter. I don't like wrapper functions that can be implemented via default parameters, because those cost when inlining isn't enabled.

The 80 character soft limit is a *soft* limit, and you can also put that parameter on a new line if you are concerned.

schveiguy commented:
BTW, I like the PR.

McSherry commented:
I've adjusted it to be a default parameter.

schveiguy commented:
LGTM, can you squash the commits?

McSherry commented:
Commits squashed.

schveiguy commented:
I just remembered, `to` wraps `parse`. Can we get the same treatment for `to`?

McSherry commented:
That should be it added for `to`.

schveiguy commented:
> That should be it added for to.

Need radix versions too :)

schveiguy commented on std/conv.d:
```diff
@@ -1738,6 +1754,18 @@ T toImpl(T, S)(S value)
     return parse!T(value);
 }
 
+T toImpl(T, S)(S value, AllowUnderscores uscores)
```
I think you need similar template constraints as above (e.g. `isDynamicArray!S`)

schveiguy replied:
Also, make sure you ddoc ditto this, or the ditto below won't work.

McSherry commented:
Oh whoops, sorry about that. _That_ should be it.

schveiguy commented on std/conv.d:
```diff
@@ -1935,7 +1969,12 @@ unittest
+alias AllowUnderscores = Flag!"AllowUnderscores";
+
+Target parse(Target, Source)(ref Source s, 
+                             AllowUnderscores uscores = AllowUnderscores.no)
```
ddoc?

schveiguy commented:
Changes for `to` look good.

McSherry commented:
One day I'll get this right on the first (or second/third) try. :smile:

Added documentation.

schveiguy commented on std/conv.d:
```diff
@@ -1739,7 +1749,22 @@ T toImpl(T, S)(S value)
-T toImpl(T, S)(S value, uint radix)
+T toImpl(T, S)(S value, AllowUnderscores uscores)
+    if ( isExactSomeString!S && isDynamicArray!S &&
+        !isExactSomeString!T && is(typeof(parse!T(value))))
```
Hm... probably want that last template constraint to be `is(typeof(parse!T(value, AllowUnderscores.no)))`

schveiguy commented on std/conv.d:
```diff
@@ -1739,7 +1749,22 @@ T toImpl(T, S)(S value)
+/// ditto
+T toImpl(T, S)(S value, uint radix, AllowUnderscores uscores = AllowUnderscores.no)
     if ( isExactSomeString!S && isDynamicArray!S &&
         !isExactSomeString!T && is(typeof(parse!T(value, radix))))
```
Similar thing here.

McSherry commented:
Revised template constraint.

schveiguy commented:
Nice! I think this is pretty useable, LGTM. At some point, we should add parsing of underscores to floating points, for consistency.

CyberShadow commented:
Um... does this really need to be in Phobos?

Does std.conv.parse accept ranges? If no, it should be fixed to accept ranges. Then just use `filter!(c => c!='_')`.

CyberShadow commented:
I don't understand how there can be consensus that this is a good change.

* Underscores are just one possible digit separator. Commas, spaces etc. are also used as digit separators in different locales.
* The problem with optional parameters is that you can realistically only can add *one* before things become messy. This feature is not "the one".
* Instead of tacking on bloat and featuritis to existing functions, we need to search for ways to combine existing features for a net gain.

schveiguy commented:
I don't think you can do that with parse. And underscores as separators are part of D. This is so one can effectively use a runtime string as input to converting to an int in the same way you can do a compile time string

CyberShadow commented:
> I don't think you can do that with parse.

Challenge accepted.

```
import std.algorithm.iteration;
import std.conv;
import std.range;

T parseWithoutUnderscores(T, S)(ref S s)
{
	auto r = refRange(&s).filter!(c => c!='_');
	return r.parse!T();
}

unittest
{
	string s;
	s = "10"   ; assert(parseWithoutUnderscores!int(s) ==    10); assert(s == "");
	s = "1_000"; assert(parseWithoutUnderscores!int(s) == 1_000); assert(s == "");
	s = "123Q" ; assert(parseWithoutUnderscores!int(s) ==   123); assert(s == "Q");
}
```

CyberShadow commented:
> This is so one can effectively use a runtime string as input to converting to an int in the same way you can do a compile time string

So, this works right now:

```d
unittest
{
	import std.conv;
	import std.stdio;
	assert(to!(string[])(`["foo", "bar"]`) == ["foo", "bar"]);
}
```

Which is pretty cool. IIRC this feature was nicknamed DSON parsing.

I understand that the goal is to also make this work:

```d
	assert(to!(int[])(`[5, 1_000]`) == [5, 1_000]);
```

I agree, this would be good.

However, we didn't get the first example to work by tacking on an extra flag to `std.conv.parse!string(string)`. I suggest to research on how to make the second example work without changing the non-DSON API.

schveiguy commented:
The reason to add the parameter is to avoid breaking code that expects e.g. parse to stop at a `_`. I haven't looked at the change that got the string[] parsing to work, but I'm guessing there wasn't a valid use case for calling to/parse on that type of string.

I agree the DSON parsing is a good motivation. I don't know that we can do it without changing the API if we want to avoid breaking existing code. If I were to choose, I'd have just included ignoring `_` in the first place.

The idea to use filter and refRange (I didn't know about that) works. It's a bit verbose, but it would work. We could potentially create a range that removed the separators, which may be a bit smarter than just blindly removing all of them.

I'm not sure what the right path is.

CyberShadow commented:
> The reason to add the parameter is to avoid breaking code that expects e.g. parse to stop at a _.

Yes. I know that, and what I said above takes this into account, so it doesn't change what I said above.

> We could potentially create a range that removed the separators, which may be a bit smarter than just blindly removing all of them.

I don't know what you mean here.

> I'm not sure what the right path is.

I think it's pretty clear to me:

- In situations where it is *unambiguous* that we want to parse DSON, e.g. `string[]` or `int[]`, there is no conflict - just add underscore parsing.
- If we want to be able to have an API for parsing DSON values where it will conflict with parsing regular values, e.g. with `parse!int`, then I think a separate API should be used, e.g. `parseD` alongside `parse`. It might already exist.
- Concentrate on semantics, not mechanics. `AllowUnderscores` is critically low in semantic meaning. Something `ParseDSyntax` is much better - but, again, optional parameters are a crappy solution to avoid API breakage - aside from the exponential "overload overload" problem, it does break cases such as `void foo(int function(string)); foo(&parse!int);`.

schveiguy commented:
OK, that sounds like a good plan.

CyberShadow commented:
> I understand that the goal is to also make this work:
> 
>      assert(to!(int[])(`[5, 1_000]`) == [5, 1_000]);
> 
> I agree, this would be good.

You know, I thought about this some more, and now I'm not sure. Because - where do we stop?

Should we also add support for parsing the other string literal syntaxes? Radix prefixes in numbers (`to!int("0x123")`)? Implicit string concatenation? You can extend this line of thought all the way to a full D interpreter, so where do you draw the line?

I think there is a well-defined line (or goal) for "DSON" right now - parse anything that you can get out of `to!string`. Underscores in numbers etc. are outside that.

So... @McSherry, what is your use case? Is it something that really should be in Phobos, and has considerable advantages over using `filter`?

schveiguy commented:
I thought radix prefixes were already processed. IMO, this should at least be as useful as `strtol`, and I think it makes sense to process strings that look like D number literals. By far, one of the most useful elements of D grammar is the underscore in number literals.

> where do you draw the line

We can draw an arbitrary line. JSON does.

I think what we need is a `parseDInteger` function that will parse a D literal formatted string. Use `filter` and `parse` if you wish under the hood. We can reuse this implementation for `octal` as well, saving some code.

If we want to then support parsing `to!(int[])("[0x123, 5_000L]")`, we have the tools to do so.

McSherry commented:
Apologies for the delayed response, I had to wait for a keyboard to be shipped.

I think that the use of a `parseD`/`parseDInteger` function would be a good place for the functionality to be (if it is to be included).

quickfur commented:
Yeah, I don't think it's a good idea to mix parsing of D literals with parsing of what a user considers as a number. Remember that D programs may be applications used by people with no idea that the program is written in D, let alone the quirks of D literal syntax. D programs written for D users is a relatively rare case in real-world applications. So parsing functions should not arbitrarily accept D-style literals where the application expects only "normal" user input.

Having a `parseD` function is a good idea -- it allows us to provide D literal parsing (or whatever other aspect of D syntax that might be useful to support) without conflating it with the general parsing functions for generic input.

quickfur commented:
ping @McSherry 
Merge conflict, please rebase.

McSherry commented:
I'll be honest, I forgot that this pull request was open. I don't really have the time/inclination to do anything about it at the moment, so I'll close it.

schveiguy commented:
@McSherry thanks for the attempt. I think it is a good idea to have this in phobos (I may attempt to resurrect this at some point), but I understand all too well how things like this can be difficult to keep up motivation for.

schveiguy commented:
Filed an issue so this isn't forgotten: https://issues.dlang.org/show_bug.cgi?id=15526
